### ISSUE TITLE
Navigator 2.0: Popping a route programmatically leaves the popped route in the browser history

### ISSUE THREAD

**Issue by carloshwa** [flutter/flutter#70588]
I've been trying to wrap my head around the Nested Navigator example for Navigator 2.0 and what the purpose of it is and why it needs to be so complex:
https://gist.github.com/johnpryan/bbca91e23bbb4d39247fa922533be7c9

I thought the main purpose was to allow programmatically 'popping' a route, but even in this example, that doesn't seem to work as expected.

## Steps to Reproduce

See this video:
https://drive.google.com/file/d/1nmXepYk3XUYCHpx_LBbGnfTU-7x2JEDU/view?usp=sharing

After tapping the `BookDetailScreen`'s back button, which calls `Navigator.of(context).pop()`:
https://gist.github.com/johnpryan/bbca91e23bbb4d39247fa922533be7c9#file-main-dart-L374

if I tap the browser's back button, it returns to the previously popped `BookDetailScreen`. Since this route was already popped, I wouldn't expect that to happen.

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel dev, 1.24.0-7.0.pre, on Mac OS X 10.15.7 19H15 darwin-x64, locale en)
    • Flutter version 1.24.0-7.0.pre at /Users/panda/pandamonium/flutter
    • Framework revision a0860f6e87 (2 weeks ago), 2020-10-29 20:07:34 -0700
    • Engine revision 073263e39d
    • Dart version 2.11.0 (build 2.11.0-260.0.dev)
```

</details>

**Comment by TahaTesser** [comment]
Hi @carloshwa 
Can you simplify your code sample to better address the issue,  provide a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example)?
Thank you

**Comment by carloshwa** [comment]
@TahaTesser Thanks for looking. The gist I linked is the 'Nested router sample' provided in the Navigator 2.0 Medium post by the Flutter team:
https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade

Since the code is part of the documentation for this component, do you still need me to provide a project with the code?

**Comment by TahaTesser** [comment]
Hi @carloshwa 
Thanks for the source, I can reproduce the issue
<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.24.0-8.0.pre.268, on macOS 11.0.1 20B29
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-8.0.pre.268 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision e803b13f01 (2 hours ago), 2020-11-16 22:55:59 -0800
    • Engine revision f0e80100f7
    • Dart version 2.12.0 (build 2.12.0-49.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.2
    • macOS (desktop)        • macos                     • darwin-x64     •
      macOS 11.0.1 20B29 darwin-x64
    • Web Server (web)       • web-server                • web-javascript •
      Flutter Tools
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 86.0.4240.198

• No issues found!
```

</details>

**Comment by carloshwa** [comment]
Perhaps this is something better covered as an improvement with #69315, but it would be nice if this ability could be achieved without having to implement a nested navigator. Just in the sample from the Medium article, the complexity of the implementation seems to grow exponentially.

It would be nice if calling `Navigator.of(context).canPop()` returns whether the Router has a route that can pop, and `Navigator.of(context).pop()` just pops the current route.

**Comment by mdebbar** [comment]
This is by design. Any route changes in your flutter app will cause a history push in the browser. This is how gmail and many other web apps behave, so we designed it around that model.

cc @chunhtai @johnpryan

**Comment by chunhtai** [comment]
It seems like what you actually want is a browser back instead of programmatically pop.
Imagine this case. you go from page A to page B, the browser also remember pageA -> pageB
What should we expect the browser history be like after a programmatically pop?

if you think browser should go back to pageA, then it is a actually browser back, and can be achieved with plugin that calls history.go(-1).

If you think browser should remember this and create this history pageA -> pageB -> pageA, then this is what programmatically pop currently does.

**Comment by carloshwa** [comment]
@mdebbar Thanks for the info. You're right, gmail behaves the same way which I had never noticed before. Well this is not the first time I've sacrificed good UX design in order to follow Google's best practices ;)

When this same sample is run on Android, the device back button behaves as one would expect: After navigating to a book detail page, then tapping the app's back button, tapping the device back button exits the app (NOT return the user back to the previous book detail page).

I assumed the Android device back button generally behaves as a browser back button would, so it makes sense to me for web to behave the same way as Android here, despite gmail's web UX.

I'm curious what the purpose of the complex nested navigator example is if this is behaving as designed?

@chunhtai Yes, performing a browser back is essentially all I need, but I cannot blindly perform a browser back, in case the user loaded the detail screen directly via URL. For example, the user enters https://localhost/#/book/1 into the browser address bar. The app's main page is not in the browser's history, so we would not be able to perform a browser back to take the user up to the main page.

In summary, in my opinion, it would be awesome if the browser back button behaved the same as the Android device button in this scenario, but I understand Flutter's decision for the current behavior given Google's UX on their own websites.

**Comment by esDotDev** [comment]
It took a while for me to understand this as well, but basically the Web "back" is not a back at all. The web is simply navigating through a set of bookmarks. The fact "back" moves one index back in the bookmark stack makes it feel like it's the same as Android, but actually there's no relationship there. 

"Back" is a specific event that occurs only on Android and iOS (I think?). 

I think this logically does make sense, but it's definitely confusing from a developer perspective, where just think of things from a cross platform perspective, and really don't want to be concerned with this distinction.

Basically if someone deeplinks into your web app (which is not possible on Android), you need to provide your own "back" or "up" btn, if what you want to do is proceed upwards. Then your normal app navigation logic should handle it.

**Comment by venkatd** [comment]
I will second what @esDotDev said. From a UX perspective, there are different expectations from a web app and from a mobile app. The back button on mobile and the back button in your browser are *not* the same. It's not a Google decision. It's what users are used to with web browsers vs. mobile apps.

Web: navigate to the link you were at before
Mobile: pop the page off the stack

They can be modified to do the same thing, but the behavior would be confusing to the user. I think it's worth documenting this UX expectation in Navigator 2.0 because it's not obvious from a dev perspective right away that there is a distinction.

For example. If you paste a url into your browser, you do not expect the history back button to work because that's your first page. However, if you click on a deep link, you might be on the top-most page in a stack of pages. In this scenario, hitting back would pop one item from the stack.

@chunhtai if I understand right, the mobile back button, the browser back button, the history, the stack of pages, can all be independently customized? If so, do you have a good place to look on how these all fit together? I have the gist of it (our app is running of Navigator 2.0) but I'm hitting into some edge cases where the browser URL isn't updating and I'm having trouble understanding the exact order in which things happen.

**Comment by chunhtai** [comment]
You can customize the stack of pages, mobile back button. The the browser back button cannot be customized, this is true for native web app as well.

The navigator 2.0 generates the browser history automatically, and we don't provide an api to customize it. You will have to write a web plugin to do so.

**Comment by esDotDev** [comment]
Why is a web plugin needed? Can't we basically do any customization we need using just `SystemNavigator.routeUpdated` and `WidgetsBinding.instance.window.defaultRouteName`? That seems to generate history just fine, and allow us to deeplink. With those 2 API's a developer can implement any implementation they like on Web... I think?

**Comment by chunhtai** [comment]
I think you mean `SystemNavigator.routeInformationUpdated`. This method pushes a browser history entry every time it is called, but there isn't much else you can do about the browser history entry. For example you cannot programatically call backward and forward, and you cannot replace current state or capture onpopstate call.

We didn't expose those API in flutter because it is web specific, and it is not need for common use case.

**Comment by esDotDev** [comment]
I have a spike running that seems to work fine using `SystemNavigator.routeUpdated`, so not really sure the difference, but I'll try switching over. You can see it running here, seems fine?? 
https://gskinnerteam.github.io/flutter-navigator-routing-examples/#/

**Comment by chunhtai** [comment]
@esDotDev The forward button doesn't work, and the backward button will just be a navigator.pop.

**Comment by carloshwa** [comment]
The points made above about web browser back button behavior make sense to me, so the current behavior is working as intended w.r.t. this issue. But the question remains: what is the purpose of the nested navigator in the medium article? It's so incredibly complicated and confusing, and I don't see what behavior it buys that a single router doesn't provide.

**Comment by chunhtai** [comment]
the nested routers may not make sense for the the use case in the blog post. but for a more complex project that require multiple teams' effort, it may start to make sense. For example if you have different teams owns different pages, they may want to have their own sub route parsing in their own page. The example nested routers in the medium post is just an example way of doing it

**Comment by esDotDev** [comment]
I switch to `routeInformationUpdated` but now am not getting any callback when the route changes. Seems `RouteInformationProvider` handles this, but I can't quite figure out how... 

Interestingly, the behavior I have with this other approach, is pretty compelling. It acts much more like an app than a classic website, occupying just 1 slot in the history, it lets you go back internally until you reach the bottom, and then it goes back to the url before it. It fully supports deeplink.

The only thing we lose here is web history  and forward support, but given how dead simple this implementation is, and how no other platforms have these concepts of Forward/History, I think it would be pretty popular even if we can't making history work easily as it really consolidates behavior across devices, and doesn't require devs to think at  worry about the different paradigm on web (web acts more like "just another platform")
